File: src/sliding-windows/longest_substring_with_unique_char.go
package slidingwindows

func LongestSubstringWithUniqueChar(s string) int {
	runes := []rune(s)
	seen := make(map[rune]bool)
	left, maxLen := 0, 0

	for right, ch := range runes {
		for seen[ch] {
			delete(seen, runes[left])
			left++
		}

		seen[ch] = true
		if length := right - left + 1; length > maxLen {
			maxLen = length
		}
	}

	return maxLen
}

File: src/sliding-windows/longest-substring-with-unique-char.ts
export function longestSubstringWithUniqueChar(s: string): number {
    const seen = new Set<string>()
    let left = 0, maxLen = 0

    for (let right = 0; right < s.length; right++) {
        // if duplicate at `right`, shrink window from `left`
        while (seen.has(s[right])) {
            seen.delete(s[left])
            left++
        }
        // now it's safe to add s[right]
        seen.add(s[right])
        maxLen = Math.max(maxLen, right - left + 1)
    }

    return maxLen
}
